### PR TITLE
Fix AmpAdApiHandler type check

### PR DESCRIPTION
**Pull request by zhouyx**
AmpAdApiHandler is calling `AmpAd3PImpl` or `AmpA4A` method, however `BaseElement` don't have these method type so there will be type check errors. This is to fix the issue by restricting the `AmpAdApiHandler` `baseInstance` to be either `AmpAd3pImpl` or `AmpA4`. This also helps for code sharing #5356 later, so that I can access their `UIHandler` without type check errors. 
PTAL

### CODE REVIEW

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 1]
```diff
@@ -32,12 +32,13 @@ const TIMEOUT_VALUE = 10000;
...
   /**
-   * @param {!AMP.BaseElement} baseInstance
+   * @param {!./amp-ad-3p-impl.AmpAd3PImpl|../../amp-a4a/0.1/amp-a4a.AmpA4A} baseInstance
    * @param {!Element} element
    * @param {function()=} opt_noContentCallback
    */
   constructor(baseInstance, element, opt_noContentCallback) {
-    /** @private {!AMP.BaseElement} */
+
+    /** @private {!./amp-ad-3p-impl.AmpAd3PImpl|../../amp-a4a/0.1/amp-a4a.AmpA4A}*/
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: The duplicated type hint here shouldn't be necessary.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 2]
```diff
@@ -32,12 +32,13 @@ const TIMEOUT_VALUE = 10000;
 export class AmpAdApiHandler {
 
   /**
-   * @param {!AMP.BaseElement} baseInstance
+   * @param {!./amp-ad-3p-impl.AmpAd3PImpl|../../amp-a4a/0.1/amp-a4a.AmpA4A} baseInstance
```
Comment (aghassemi) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: should this be {!...|!...} ? (note the second ! for the second union type)

Comment (aghassemi) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: (same for the instance variable one)

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 1 above)
Comment (aghassemi) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: I think we need this. Even if CC managed to infer the type from the param, without a `@private` it won't consider it an instance variable.

Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: yes, i think `this.baseInstance_` still need type check, like  we still put a `@private {!Element} before `this.element_`.

Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: `@private`'s not a type hint, we need it. But the `{...}` stuff should be inferred, right?

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -161,9 +161,10 @@ export class AmpA4A extends AMP.BaseElement {
     /** @private {?string} */
     this.experimentalNonAmpCreativeRenderMethod_ = null;
 
-    this.lifecycleReporter_ = getLifecycleReporter(this, 'a4a');
+    /** @public {!../../../ads/google/a4a/performance.AmpAdLifecycleReporter|!../../../ads/google/a4a/performance.NullLifecycleReporter} */
```
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: `@public` isn't needed.

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 4]
```diff
@@ -84,8 +84,10 @@ export class AmpAd3PImpl extends AMP.BaseElement {
     /** @private {?Promise} */
     this.layoutPromise_ = null;
 
-    this.lifecycleReporter_ = getLifecycleReporter(this, 'amp');
-    this.lifecycleReporter_.sendPing('adSlotBuilt');
+    /** @public {!../../../ads/google/a4a/performance.AmpAdLifecycleReporter|!../../../ads/google/a4a/performance.NullLifecycleReporter} */
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: ditto.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 1 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: I see. it did work with only `@private`. Hmmm the type check thing confused me so much 😢

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (zhouyx) [inline extensions/amp-a4a/0.1/amp-a4a.js]: fixed

**Pull request merged**